core/repository/http/src/main/java/org/eclipse/rdf4j/repository/http/HTTPTupleQuery.java

```diff
@@ -48,8 +48,11 @@ public TupleQueryResult evaluate() throws QueryEvaluationException {
 		SPARQLProtocolSession client = getHttpClient();
 		try {
 			conn.flushTransactionState(Protocol.Action.QUERY);
+
+			WeakReference<HTTPTupleQuery> callerRef = new WeakReference<>(this);
+
 			return client.sendTupleQuery(queryLanguage, queryString, baseURI, dataset, getIncludeInferred(),
-					getMaxExecutionTime(), new WeakReference<>(this), getBindingsArray());
+					getMaxExecutionTime(), callerRef, getBindingsArray());
 		} catch (IOException | RepositoryException | MalformedQueryException e) {
 			throw new HTTPQueryEvaluationException(e.getMessage(), e);
 		}
```

Comment:
@JervenBolleman

Ontotext is experiencing some issues with FedX queries being cancelled before they are done. They have narrowed it down to the introduction of weak references which were intended to close iterations running in a different thread (background) if the caller had forgotten to close the query result. 

I've expanded the code a bit here by introducing a variable for the callerRef. I believe that the core of the issue is that the weak reference wraps the tuple query object, resulting in the iteration being closed if the tuple query is garbage collected.

The code below could potentially see the tuple query object being garbage collected before the tuple query results were fully consumed.


```
	public static void main(String[] args) {
		try (RepositoryConnection connection = repo.getConnection()) {
			try (TupleQueryResult evaluate = getEvaluate(connection)) {
				for (BindingSet bindings : evaluate) {
					System.out.println(bindings);
				}
			}
		}
	}

	private static TupleQueryResult getEvaluate(RepositoryConnection connection) {
		TupleQuery tupleQuery = connection.prepareTupleQuery("select * where {?a ?b ?c}");
		return tupleQuery.evaluate();
	}
```

Reply:
Yes, I see the problem and this is a better solution. Still fragile as depending on cleaner tasks. It was a missed opportunity to really fix this for 4.x with a shared variable ;( only so much time in a day.